x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/profile/ProfileService.java

```diff
@@ -344,18 +350,101 @@ private void createNewProfile(Subject subject, ActionListener<Profile> listener)
-                }, listener::onFailure))
+                }, e -> {
+                    if (e instanceof VersionConflictEngineException) {
+                        // Document already exists with the specified ID, get the document with the ID
+                        // and check whether it is the right profile
+                        getVersionedDocument(profileDocument.uid(), ActionListener.wrap(versionedDocument -> {
+                            if (isTheRightProfileDocument(subject, versionedDocument.doc)) {
+                                // The profile document is a match, still need to update it with current auth info before return
+                                // listener.onResponse(versionedDocument.toProfile(Set.of()));
+                                updateProfileForActivate(subject, versionedDocument, listener);
+                            } else {
+                                // The profile document is NOT a match, this means either genuine hash collision or profile document
+                                // was manually updated, e.g. realm or user rename etc. So we attempt to differentiate from the
+                                // existing profile document by increase the differentiator number by 1.
+                                incrementDifferentiatorAndCreateNewProfile(subject, profileDocument, listener);
+                            }
+                        }, listener::onFailure));
+                    } else {
+                        listener.onFailure(e);
+                    }
+                }))
             )
         );
     }
 
+    private void incrementDifferentiatorAndCreateNewProfile(
+        Subject subject,
+        ProfileDocument profileDocument,
+        ActionListener<Profile> listener
+    ) throws IOException {
+        final String profileUid = profileDocument.uid();
+        final int index = profileUid.lastIndexOf("_");
+        if (index == -1) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] does not contain any underscore character", profileUid));
+            return;
+        }
+        final String differentiator = profileUid.substring(index + 1);
+        if (differentiator.isBlank()) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] does not contain a differentiator", profileUid));
+            return;
+        }
+        final int differentiatorNumber;
+        try {
+            differentiatorNumber = Integer.parseInt(differentiator);
+        } catch (NumberFormatException e) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator is not a number", e, profileUid));
+            return;
+        }
+        // TODO: temporary to prevent infinite recursion
+        if (differentiatorNumber > 99) {
+            listener.onFailure(new ElasticsearchException("profile uid [{}] differentiator number is too high (>99)", profileUid));
+            return;
+        }
```

Comment:
I don't think it could be a real scenario for having such high number of hash collision. But it is still nice to proactively prevent infinite recursion. This is just temporary. We can do better. One idea is to get all profile documents that has UID beginning with the clashed base uid and figure out the actual increment number instead of always incrementing by one in each recursion.